File: leetcode-solutions/binary-search/solution-1231/solution.go
package solution1231

import (
    "math"
)

func maximizeSweetness(sweetness []int, k int) (res int) {
    l := 0
    r := int(math.Pow10(9))
    for l < r {
        m := int(math.Ceil(float64(l+r) / 2))
        count := 0
        cur := 0
        min := math.Inf(1)
        for _, v := range sweetness {
            cur += v
            if cur >= m {
                count++
                min = math.Min(min, float64(cur))
                cur = 0
            }
        }
        if count >= k+1 {
            l = m
            res = int(min)
        } else {
            r = m - 1
        }
    }
    return
}

File: leetcode-solutions/binary-search/solution-1231/solution.js
/**
 * @param {number[]} sweetness
 * @param {number} k
 * @return {number}
 */
var maximizeSweetness = function(sweetness, k) {
  let l = 0;
  let r = 10**9;
  let res;
  while (l < r) {
    let m = Math.ceil((l + r) / 2);
    let count = 0;
    let cur = 0;
    let min = Infinity;
    for (let i = 0; i < sweetness.length; i++) {
      cur += sweetness[i];
      if (cur >= m) {
        min = Math.min(min, cur);
        count++;
        cur = 0;
      }
    }
    if (count >= k + 1) {
      l = m;
      res = min;
    } else {
      r = m - 1;
    }
  }
  return res;
};
